src/parsers/gene_association_file.py

```diff
@@ -0,0 +1,117 @@
+    """
+    Processes GO annotations by cleaning, transforming, and mapping evidence types.
+
+    Args:
+        input_path (str): Path to the input GO annotations CSV file.
+        output_path (str): Path to save the normalized output CSV file.
+    """
+    
+    # load the input data 
+    df = pd.read_csv(input_path)
+
+    # Drop unnecessary columns to simplify the DataFrame
+    drop_columns = [
+        "DB_Object_Symbol", "Aspect", "DB_Object_Name", "Synonym", 
+        "DB_Object_Type", "Taxon", "Annotation_Extension", "Gene_Product_Form_ID"
+        ]
+    
+    df = df.drop(columns = drop_columns)
+
+    # Create a unified 'subject' identifier
+    df["subject"] = df["DB"] + ":" + df["DB_Object_ID"]
+    df = df.drop(columns=["DB", "DB_Object_ID"])
+
+    # Reorder columns to place 'subject' first
+    cols = ["subject"] + [col for col in df.columns if col != "subject"]
+    df = df[cols]
+    
+    # Rename columns to match the desired schema
+    df = df.rename(columns={
+        "Qualifier": "predicate", 
+        "GO_ID": "object",
+        "DB_Reference": "publications", 
+        "With_From": "supporting_objects",
+        "Date": "annotation_date",
+        "Assigned_By": "primary_knowledge_source"})
+    
+    # Convert 'annotation_date' to a standard format
+    df["annotation_date"] = pd.to_datetime(df["annotation_date"], format='%Y%m%d').dt.strftime('%Y-%m-%d')
+    df["annotation_date"] = df["annotation_date"].astype(str)
+    
+    # Add constant metadata columns 
+    df["aggregator"] = "UniProt"
+    df["protocol_id"] = "NULL"
+    df["supporting_objects"] = df["supporting_objects"].fillna("NULL")
+    
+    # Define allowed predicates, determine negation, and clean predicate names
+    allowed_predicates = ["enables", "contributes_to", "acts_upstream_of_or_within", "involved_in",
+                          "acts_upstream_of", "acts_upstream_of_positive_effect",
+                          "acts_upstream_of_negative_effect", "acts_upstream_of_or_within_negative_effect",
+                          "acts_upstream_of_or_within_positive_effect", "located_in", "part_of", 
+                          "is_active_in", "colocalizes_with"]
+
+    df["negated"] = ~df["predicate"].isin(allowed_predicates)
+    df["predicate"] = df["predicate"].str.replace(r"^NOT\|", "", regex=True)
+    
+    # Load the external ECO mapping data
+    url = "https://raw.githubusercontent.com/evidenceontology/evidenceontology/master/gaf-eco-mapping.txt"
+    mapping_columns = ["Evidence_Code", "DB_Reference", "evidence_type"]
+    df2 = pd.read_csv(url, sep="\t", comment="#", header=None, names=mapping_columns)
+    
+    """
+    merge two daraframe
+    df1: go annotations data
+    df2: ECO mapping data
+    df2 DB_reference match with df1 publications
+    df2 evidence code match with df1 evidence code 
+    """
+
+    # Normalize strings for consistent matching 
+    df["publications"] = df["publications"].str.strip().str.upper()
```

Comment:
This col is going to need a bit more attention as it sometimes contains more than one entry. [The docs aren't 100% clear on what this col contains](https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/#dbreference-column-6) so let's talk about this later on. :)